src/nn/Compilation.ts

```diff
@@ -0,0 +1,68 @@
+import { CompilationOptions } from "./CompilationOptions";
+import { Execution } from "./Execution";
+import { ExecutionContext } from "./ExecutionContext";
+import { Model } from "./Model";
+import { Constant } from "./Constant";
+import { Input } from './Input';
+import { getTypedArray, sizeFromDimensions, createTensor, createOperandDescriptorFromTensor } from "./utils";
+import { Output } from "./Output";
+import { OperandDescriptor } from "./OperandDescriptor";
+
+import * as tf from '@tensorflow/tfjs-core'
+import "@tensorflow/tfjs-backend-webgl";
+import '@tensorflow/tfjs-backend-cpu';
+
+export class Compilation {
+  model_: Model;
+  constantTensors_: Map<Constant, tf.Tensor> = new Map();
+  outputDescriptors_: Map<Output, OperandDescriptor> = new Map();
+
+  constructor(options: CompilationOptions, model: Model) {
+    // TODO: support compilation options.
+    this.model_ = model;
+  }
+
+  async compile(): Promise<void> {
+    if (!(await tf.setBackend('webgl'))) {
+      console.warn('Failed to set tf.js webgl backend, fallback to cpu backend.');
+      if (!(await tf.setBackend('cpu'))) {
+        throw new Error('Failed to set tf.js cpu backend.');
+      }
+    }
+    await tf.ready();
+    this.allocateConstants_();
+    await this.inferOutputShapes_();
+  }
+
+  allocateConstants_ () {
+    for (const constant of this.model_.constants_) {
+      this.constantTensors_.set(constant, constant.createTensor());
+    }
+  }
+
+  async inferOutputShapes_() {
+    const inputTensors:  Map<Input, tf.Tensor> = new Map();
+    for (const input of this.model_.inputs_.values()) {
+      const typedArrayConstructor = getTypedArray(input.desc.type);
+      const inputBuffer = new typedArrayConstructor(sizeFromDimensions(input.desc.dimensions));
+      inputTensors.set(input, createTensor(input.desc, inputBuffer));
+    }
+    for (const output of this.model_.outputs_.values()) {
+      const tensor: tf.Tensor = tf.tidy(() => {
+        return output.operation.run({
```

Comment:
Does this trigger an inference call on the model?

Reply:
can this include async ops? meaning its execution depends on the value of the input tensors.

Reply:
> Does this trigger an inference call on the model?

Yes.

> can this include async ops? meaning its execution depends on the value of the input tensors.

Yes. I create empty input tensors for it.

Reply:
Does it look good to you? @pyu10055

Reply:
lgtm, thanks

> > * Are there thoughts on how to execute sub-graph?
> 
> I suppose each sub-graph can be executed by corresponding `Model`, `Compilation` and `Execution` objects respectively. I am not sure I got your context correctly. What's the exact usage of the sub-graph? @pyu10055

subgraph execution is when you need to execute part of the model, for example you want to do transfer learning, you would execute the model to the feature vector node, not the final classification node.

Reply:
this can come later, I just want to raise some awareness on this.

Reply:
sounds good, thanks for pointing this out.